### PR TITLE
Add support for Java21 string template

### PR DESCRIPTION
**Pull request by butterunderflow**
86390d0d569d5847d00c3938ec60ad444a3c39e7: Support for string templates and corresponding tests. But won't work, because output tokens of `com.sun.tools.javac.parser.Scanner` are not sorted by their position, and then cause the verification [here](https://github.com/google/google-java-format/blob/4c1aeffc8b0ea3d4e18ffe21162b835f2c51cc3d/core/src/main/java/com/google/googlejavaformat/OpsBuilder.java#L328) failed. 
86390d0d569d5847d00c3938ec60ad444a3c39e7: Read all tokens from Scanner, and sort them if jdk version >= 21.

Some links may be useful:
How jdk parse string template from tokens: https://github.com/openjdk/jdk/blob/3d9d353edb64dd364925481d7b7c8822beeaa117/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java#L695-L746
How tokens of string templates are builded: https://github.com/openjdk/jdk/blob/3d9d353edb64dd364925481d7b7c8822beeaa117/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavaTokenizer.java#L342-L420

### CODE REVIEW

**Review by cushon — CHANGES_REQUESTED** [review]
Thanks for this!

I had started working on this in parallel before seeing this PR, and had merged a very simple initial handling of string templates in https://github.com/google/google-java-format/commit/b5feefe4ab4dbdbba393d769eb1815f26896104b. It just passes the entire string template through without formatting it. This approach is more complete, and I would like to merge most of what you have here.

**Comment by butterunderflow** [comment]
Thank you for this detailed reviewing!

File: core/src/main/java/com/google/googlejavaformat/java/JavacTokens.java
Comment (butterunderflow) [reply core/src/main/java/com/google/googlejavaformat/java/JavacTokens.java:127]: I think your demo will work without problem. 

There's one thing I'm not very sure, will that too tricky to split `\` and `{` into two tokens? For example, can we guarantee that `{` will follow `\` immediately?

After some test of your demo, I can't find a negative test case to illustrate this. Even for very long string fragment, `\` and `{` are stay together perfectly.

**Comment by butterunderflow** [comment]
Should I close this PR and wait for yours to merge? There are some details in the original PR that weren't considered, and I found they're already handled in your fork.

File: core/src/main/java/com/google/googlejavaformat/java/JavacTokens.java
Comment (cushon) [reply core/src/main/java/com/google/googlejavaformat/java/JavacTokens.java:127]: I think that should be OK, and that `\{` is handled as an escape sequence where it wouldn't be valid to have any separate between the two characters.

**Comment by cushon** [comment]
> Should I close this PR and wait for yours to merge? There are some details in the original PR that weren't considered, and I found they're already handled in your fork.

Thanks for taking a look! If the changes I made look OK I will go ahead and merge that version (and credit you).

**Comment by cushon** [comment]
This has been merged in 38de9c4e05bc871496c232c1ccbfc4be06a580e2. Thanks again for the contribution!

**Pull request closed without merging**
